docs/extended-collection-types.md

```diff
@@ -0,0 +1,169 @@
+
+### Array
+
+A fixed-length collection of elements. Almost identical to `list()` except it cannot be resized.
+
+Example usage:
+```
+var/array/foo = array("a","b","c")
+foo += "z" // Compile-time error, cannot resize arrays
+
+world.log << foo.Join() // Prints "abc"
+
+foo[1] = "z" // Valid modification
+```
+
+### Dictionary
+A collection of key-value pairs. Equivalent to existing associative lists, and will serve as the internal implementation of BYOND's planned `alist()` feature.
+
+Example usage:
+```
+var/dict/bar = array("a" = 1,"b" = 2,"c" = 3)
+bar += "d" // Compile-time error, no value specified
+
+bar["d"] = 4 // Valid modification
+
+world.log << bar["c"] // Prints 3
+
+world.log << bar["test"] // Runtime error, key not found
+
+bar.contains_key("test") // Returns FALSE
+
+bar.get_value("test") // Returns null, since the key doesn't exist
+
+world.log << bar[1] // Compile-time error, dicts can't be numerically indexed
+```
+
+`dict.keys` and `dict.values` will return a new `array()` of the dictionary's keys and values, respectively. Modifying these will not modify the existing dictionary.
+
+`dict.contains_key(key)` will return `TRUE` or `FALSE` depending on if the `key` is present in the dictionary.
+
+`dict.get_value(key)` will return the value of the specified `key` or `null` if the key is not present in the dictionary.
+
+### HashSet
+Essentially identical to an OD list, except it cannot contain duplicates. Inserting an element that is already present in the hashset is not an error and instead does nothing.
+
+Direct insertion (e.g. `hashset += "foo"`) is supported, but `hashset.add(element)` will return `TRUE` or `FALSE` depending on if the element already existed in the hashset.
+
+Example usage:
+```
+var/hashset/foo = hashset("a","b","c")
+
+foo += "c" // Valid but nothing happens since `foo` already contains "c"
+
+foo["c"] = 5 // Compile-time error, hashsets cannot be associative
+
+world.log << foo.add("c") // Prints FALSE and inserts nothing since `foo` already contains "c"
+
+world.log << foo.add("d") // Prints TRUE and inserts "d" into `foo`
+world.log << foo.Join() // Prints "abcd"
+```
+
+### ODList 
+Identical to BYOND `list()` but with no support for associative lists. This facilitates various internal optimizations over using `list()`.
+
+Unlike BYOND lists, ODLists also implement the "Constant Collections" and "Collection Capacity API" defined in subsequent sections.
+
+Example usage:
+```
+var/odlist/foo = odlist("a","b","c")
```

Comment:
`odlist` seems rather awkward of a name. something like `dynarray` or `collection` seems a little bit more fitting.

Reply:
`collection` is too generic and `dynarray` is cringe.

The rationale here is that `odlist` prevents namespace pollution and codebases can wrap it in a macro if they want to rename it.

Reply:
Couldnt think of a better name; but that seems good albeit smidge weird.

Reply:
~~vec~~